File: r.kmeans.py
# k-means

# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd 

# Importing the dataset
dataset = pd.read_csv('Mall_Customers.csv')   
X = dataset.iloc[:, [3,4]].values

# Using the elbow method to find the optimal number of clusters
from sklearn.cluster import KMeans
wcss = [] #list wcss
for i in range(1,11):
    kmeans = KMeans(n_clusters =  i, init = 'k-means++', max_iter = 300, n_init = 10, random_state = 0)
    kmeans.fit(X)  
    wcss.append(kmeans.inertia_) 
plt.plot(range(1,11), wcss)
plt.title('The Elbow Method')
plt.xlabel('Number of Clusters')
plt.ylabel('WCSS')
plt.show()

# Applying K-means to the dataset
kmeans = KMeans(n_clusters =  5, init = 'k-means++', max_iter = 300, n_init = 10, random_state = 0)
y_kmeans = kmeans.fit_predict(X) 

# Visualizing the clusters   
plt.scatter(X[y_kmeans == 0, 0], X[y_kmeans == 0, 1], s = 100, c = 'red', label = 'Careful')  
plt.scatter(X[y_kmeans == 1, 0], X[y_kmeans == 1, 1], s = 100, c = 'blue', label = 'Standard')  
plt.scatter(X[y_kmeans == 2, 0], X[y_kmeans == 2, 1], s = 100, c = 'green', label = 'Target')  
plt.scatter(X[y_kmeans == 3, 0], X[y_kmeans == 3, 1], s = 100, c = 'cyan', label = 'Careless')  
plt.scatter(X[y_kmeans == 4, 0], X[y_kmeans == 4, 1], s = 100, c = 'magenta', label = 'Sensible')
plt.scatter(kmeans.cluster_centers_[:,0], kmeans.cluster_centers_[:,1], s = 300, c = 'yellow', label = 'Centroids') #plotting the centroids
plt.title('Clusters of customers')                                            #cluster_centers_ returns the coordinates of the centroids
plt.xlabel('Annual Income (k$)')
plt.ylabel('Spending Score (1-100)')
plt.legend()
plt.show()                                                                       

                                                         

File: r.kmeans.r
#K-Means Clustering
                               #K Means is based on distance,so we can use Scaled values as inputs to avoid 
                             #the final cluster output to be biased towards the larger valued attributes in the dataset
#Importing the dataset
dataset = read.csv('Mall_Customers.csv')
X = dataset[4:5] #array X

# Using the elbow method to find the optimal number of clusters
set.seed(6)
wcss <- vector()
for (i in 1:10) wcss[i] <- sum(kmeans(X, i)$withinss) #i clusters
plot(1:10, wcss, type = 'b', main = paste('Clusters of clients'), xlab = 'Number of clusters', ylab = 'WCSS')#b-both(points,lines)

# Fitting K-Means to the dataset
set.seed(29)
kmeans = kmeans(X, 5, iter.max = 300, nstart = 10) #centers = 5
y_kmeans = kmeans$cluster
#dataset$cluster_number <- y_kmeans 

# Visualising the clusters (scaled by k-means)
#install.packages('cluster')
library(cluster)
clusplot(X,
         kmeans$cluster,
         line = 0,
         shade = TRUE,
         color = TRUE,
         labels = 2,
         plotchar = FALSE,
         span = TRUE,
         main = paste('Clusters of Clients'),
         xlab = 'Annual Income',
         ylab = 'Spending Score')

#to visualise the clusters without scaling factor(actual clusters)
plot(x= X[,1], y= X[,2], col= y_kmeans, pch=19, 
     xlim=c(from=min(dataset[,1]), to=max(dataset[,1]+30)),
     xlab="Annual Income", ylab="Spending Score")
clusters=c("Careless", "Standard", "Sensible", "Target", "Careful")
legend('bottomright', legend=clusters, col=1:5, pch=19, horiz=F)